rust/wal3/src/shard_manager.rs

```diff
@@ -0,0 +1,221 @@
+}
+
+/////////////////////////////////////////// ManagerState ///////////////////////////////////////////
+
+#[derive(Debug)]
+#[allow(clippy::type_complexity)]
+struct ManagerState {
+    shards: Vec<ShardState>,
+    enqueued: Vec<(
+        StreamID,
+        Message,
+        tokio::sync::oneshot::Sender<Result<LogPosition, Error>>,
+    )>,
+    last_batch: Instant,
+}
+
+impl ManagerState {
+    fn select_shard_for_write(
+        &mut self,
+        options: &ThrottleOptions,
+        manifest_manager: &ManifestManager,
+        record_count: usize,
+    ) -> Option<(ShardID, ShardSeqNo, LogPosition, DeltaSeqNo)> {
+        let shard_state = self
+            .shards
+            .iter_mut()
+            .filter(|x| x.writers_active < options.outstanding)
+            .min_by_key(|x| x.next_write)?;
+        if shard_state.next_write > Instant::now() {
+            return None;
+        }
+        let (log_position, delta_seq_no) = match manifest_manager.assign_timestamp(record_count) {
+            Some(log_position) => log_position,
+            None => {
+                panic!("log full");
+            }
+        };
+        let next_seq_no = shard_state.next_seq_no;
+        shard_state.writers_active += 1;
+        shard_state.set_next_write(options);
+        Some((
+            shard_state.shard_id,
+            next_seq_no,
+            log_position,
+            delta_seq_no,
+        ))
+    }
+
+    fn finish_write(&mut self, shard_id: ShardID) {
+        let shard_state = self
+            .shards
+            .iter_mut()
+            .find(|x| x.shard_id == shard_id)
+            .unwrap();
+        shard_state.writers_active -= 1;
+    }
+}
+
+/////////////////////////////////////////// ShardManager ///////////////////////////////////////////
+
+#[derive(Debug)]
+pub struct ShardManager {
+    options: ThrottleOptions,
+    state: Mutex<ManagerState>,
+    records_written: AtomicUsize,
+    batches_written: AtomicUsize,
+    write_finished: tokio::sync::Notify,
+}
+
+impl ShardManager {
+    pub fn new(options: ThrottleOptions, initial_manifest: &Manifest, shards: usize) -> Self {
+        let shards = (1..=shards)
+            .map(|idx| {
+                let shard_id = ShardID(idx);
+                let next_seq_no = initial_manifest
+                    .next_seq_no_for_shard(shard_id)
+                    .unwrap_or(ShardSeqNo(1));
+                let next_write = Instant::now();
+                ShardState {
+                    shard_id,
+                    next_seq_no,
+                    next_write,
+                    writers_active: 0,
+                }
+            })
+            .collect();
+        Self {
+            options,
+            state: Mutex::new(ManagerState {
+                shards,
+                enqueued: Vec::new(),
+                last_batch: Instant::now(),
+            }),
+            records_written: AtomicUsize::new(100_000),
+            batches_written: AtomicUsize::new(1),
```

Comment:
What is `records_written` and `batches_written`? (if they should be interpreted as is, why are they non-zero upon initialization)

Reply:
I've left a comment.  Cheap hack to divide-by-zero and open up the window fast.